src/Backups/BackupIO_S3.cpp

```diff
@@ -86,9 +86,10 @@ BackupReaderS3::BackupReaderS3(
     const S3::URI & s3_uri_, const String & access_key_id_, const String & secret_access_key_, const ContextPtr & context_)
     : s3_uri(s3_uri_)
     , client(makeS3Client(s3_uri_, access_key_id_, secret_access_key_, context_))
-    , max_single_read_retries(context_->getSettingsRef().s3_max_single_read_retries)
     , read_settings(context_->getReadSettings())
+    , request_settings(context_->getStorageS3Settings().getSettings(s3_uri.uri.toString()).request_settings)
 {
+    request_settings.max_single_read_retries = context_->getSettingsRef().s3_max_single_read_retries; // FIXME: Avoid taking value for endpoint
```

Comment:
Returning previous behaviour fixes tests. Problem was with `BackupReaderS3` only, but I returned writer behaviour also. Taking settings from per-endpoint storage hides values from settings. I think fixing test would be better, but I'm not sure how. Maybe @vitlibar can do this later.